### DISCUSSION TITLE
consumeNewest understanding

### DISCUSSION THREAD

**Discussion by AlexisSouquiere** [tchiotludo/akhq#1361] [General]
Hi ! I'm digging into the source code to understand some painpoint we can have currently and there is a topic I'd like to discuss.
Assuming we are using the Newest sort by default to ease topic analysis and a poll-timeout set to 15s by default I found that we will always have a response time greater than 15s to consult topic data even with topics with few data. If we leave the timeout to 1s it's ok but I'm not sure if this behaviour is expected or not

I did some tests locally and added some logs and found that this response time is caused by the do/while empty polling in the consumeNewest function. Basically we will do a 1st loop turn to get 15 records (5 records by partitions - 3 partitions in the topic) by consuming the topic from the last offset of each partition minus 5 records. Normally we should stop here because we have our 15 records to return but we will do an extra loop turn to consume data from the real end offset but because we are at the end, we will wait 15s to the timeout and then return the result to the user.

I did a picture to explain the way it works with a topic / 3 partitions and I'd like to know more about it and why the consumeNewest is directly impacted by the poll-timeout variable. IMO we can either remove the do/while loop or add a condition in the while to stop if we are at the end offset but I'd like some insights before doing a PR to check if this loop is needed elsewhere

![image](https://user-images.githubusercontent.com/2262145/216573723-279652a0-b509-4382-adef-857801ca8226.png)
 
Thanks !

**Comment by tchiotludo** [comment]
Clearly, what you described so far is a real bug, and the second loop must be the real issue and an be remove.
The consumeNewest function was really old and I have some pain to implement in my memory. I have edge case that let me poll 2 time to be sure to be at the end but clearly it was a first thought and I'm sure it can be optimized, for read tell me, just add a control on message count to avoid the second loop will do the tricks! 
Thanks for the analysis, and for sure PR are welcome :tada:

**Comment by AlexisSouquiere** [comment]
Ok thank you for the answer ! I'll work on it then and propose a PR
